##### Andre's single-cell RNA-Seq analysis
##### re-clustering using immune cells

## subsetting immune cell clusters only
```{r}
library(Seurat)

setwd("C:/Users/kicheol/Desktop/Projects/Andres_scRNA/Results_Seurat_new_NoCorrection/ImmuneCells")
setwd("C:/Users/kicheol/Desktop/Projects/Andres_scRNA/Results_Seurat_new_NoCorrection/RData")

# select cells by metadata information
immune.cells <- SubsetData(
  object = experiment.aggr,
  cells.use = rownames(experiment.aggr@meta.data)[experiment.aggr@meta.data$res.3.5 %in% c(2,3,5,7,9,12,16)])

TSNEPlot(object = immune.cells, pt.size=1, do.label = T, pt.shape="orig.ident", label.size=3.5)

save(immune.cells,file="Seurat_object_immune_original.RData")
```


## Identify variable genes

```{r}
immune.cells <- FindVariableGenes(
  object = immune.cells,
  mean.function = ExpMean,
  dispersion.function = LogVMR,
  x.low.cutoff = 0.2,
  x.high.cutoff = 5,
  y.cutoff = 0.1)

length(immune.cells@var.genes)      # select highly variable genes (reduce for PCA or else)
```

```{r}
save(immune.cells, file="Seurat_object_immune_re-variable.RData")
load(file="Seurat_object_immune_re-variable.RData")
```


make variable gene list
```{r}
var.genes <- sapply(X=immune.cells@var.genes, FUN=list)
p <- lapply(X=var.genes, FUN=function(var.gene){
  VlnPlot(object=immune.cells, features.plot = var.gene, group.by = "orig.ident")
})

p <- lapply(X=head(var.genes), FUN=function(var.gene){
  VlnPlot(object=immune.cells, features.plot = var.gene, group.by = "orig.ident")
})  # head makes it quick

head(p)
p$Fam178b

```

## Dimensionality reduction with PCA 


```{r}
immune.cells <- RunPCA(
  object = immune.cells,
  pc.genes = immune.cells@var.genes,
  do.print = TRUE,
  pcs.print = 1:3,
  genes.print = 5,
  pcs.compute = 40,
  maxit = 500)

PrintPCAParams(immune.cells)
```

Seurat then provides a number of ways to visualize the PCA results
```{r}
PCAPlot(
  object = immune.cells,
  dim.1 = 1,
  dim.2 = 2 )
```

Visualize top genes associated with principal components
```{r}
VizPCA(
  object = immune.cells,
  pcs.use=1:4
)
```

Draws a heatmap focusing on a principal component. Both cells and genes are sorted by their principal component scores. Allows for nice visualization of sources of heterogeneity in the dataset.

```{r}
PCHeatmap(
    object = immune.cells, 
    pc.use = 1:6, 
    cells.use = 500, 
    do.balanced = TRUE, 
    label.columns = FALSE,
    use.full = FALSE
)
PCHeatmap(
    object = immune.cells, 
    pc.use = 7:12, 
    cells.use = 500, 
    do.balanced = TRUE, 
    label.columns = FALSE,
    use.full = FALSE
)
```

### Selecting which PCs to use
```{r}
PCElbowPlot(
  immune.cells,
  num.pc = 40)
```


```{r}
# permutation - JackStraw
immune.cells <- JackStraw(
    object = immune.cells, 
    num.replicate = 100, 
    num.pc = 40,
    do.print = FALSE
)

JackStrawPlot(object = immune.cells, PCs = 1:40, nCol = 5)
```

Looking at the results of the JackStraw plot, we determine to use the first 35 PCs
```{r}
use.pcs = 1:8
```

## Identifying clusters
```{r}
immune.cells <- FindClusters(
    object = immune.cells, 
    reduction.type = "pca", 
    dims.use = use.pcs, 
    resolution = seq(0.5,4,0.5), 
    print.output = FALSE, 
    save.SNN = TRUE
)
PrintFindClustersParams(object = immune.cells)



sapply(grep("^res",colnames(immune.cells@meta.data),value = TRUE),
       function(x) length(unique(immune.cells@meta.data[,x])))

immune.cells <- SetAllIdent(immune.cells, id = "res.2")

immune.cells <- SetAllIdent(immune.cells, id = "res.3")
table(immune.cells@ident,immune.cells@meta.data$batchid)



immune.cells <- RunTSNE(
  object = immune.cells,
  reduction.use = "pca",
  dims.use = use.pcs,
  do.fast = TRUE)
```


Plot TSNE coloring by the slot 'ident'

```{r}
TSNEPlot(object = immune.cells, pt.shape="batchid", pt.size=1.5)
```

Plot TSNE coloring by the slot 'orig.ident' (sample names).
```{r}
TSNEPlot(object = immune.cells, group.by="batchid", pt.size=1.2)


# Plot TSNE coloring by the clustering resolution 4.5
pdf("tSNE_plots_immune.pdf", paper='USr')
TSNEPlot(object = immune.cells, group.by="batchid", plot.title="tNSE, time-point", pt.size=1.2)
TSNEPlot(object = immune.cells, group.by="res.0.5", plot.title="tNSE (res=0.5)", pt.size=1.2, do.label = TRUE)
TSNEPlot(object = immune.cells, group.by="res.1", plot.title="tNSE (res=1)", pt.size=1.2, do.label = TRUE)
TSNEPlot(object = immune.cells, group.by="res.1.5", plot.title="tNSE (res=1.5)", pt.size=1.2, do.label = TRUE)
TSNEPlot(object = immune.cells, group.by="res.2", plot.title="tNSE (res=2)", pt.size=1.2, do.label = TRUE)
TSNEPlot(object = immune.cells, group.by="res.2.5", plot.title="tNSE (res=2.5)", pt.size=1.2, do.label = TRUE)
TSNEPlot(object = immune.cells, group.by="res.3", plot.title="tNSE (res=3)", pt.size=1.2, do.label = TRUE)
TSNEPlot(object = immune.cells, group.by="res.3.5", plot.title="tNSE (res=3.5)", pt.size=1.2, do.label = TRUE)
TSNEPlot(object = immune.cells, group.by="res.4", plot.title="tNSE (res=4)", pt.size=1.2, do.label = TRUE)
dev.off()

TSNEPlot(object = immune.cells, group.by="res.0.5", pt.size=1.5, do.label = TRUE)
```

FeaturePlot can be used to color cells with a 'feature', non categorical data, like number of UMIs
```{r}
FeaturePlot(immune.cells, features.plot=c('nUMI'), pt.size=1.2, no.legend = FALSE)
# and number of genes present
FeaturePlot(immune.cells, features.plot=c('nGene'), pt.size=1.2, no.legend = FALSE)
# and percent mitochondrial 
FeaturePlot(immune.cells, features.plot=c('percent.mito'), pt.size=1.2, no.legend = FALSE)
```

## Building  a  tree relating the 'average' cell from each cluster

```{r}
immune.cells <- SetAllIdent(immune.cells, id = "res.2")
immune.cells <- BuildClusterTree(
  immune.cells,
  do.reorder = F,
  reorder.numeric = F,
#  SNN.use = immune.cells@snn,
  do.plot=F)

PlotClusterTree(immune.cells)
```

Plot the split at node 30
```{r}
ColorTSNESplit(immune.cells, node = 15)
```


## Identifying Marker Genes

```{r}
immune.cells <- SetAllIdent(immune.cells, id = "res.0.5")    # select resolution of tree
markers = FindMarkers(immune.cells, ident.1=c(0), genes.use=rownames(immune.cells@scale.data))

head(markers)
dim(markers)
table(markers$avg_logFC > 0)
```


```{r}
VlnPlot(object = immune.cells, features.plot = rownames(markers)[1:10])
```

feature plot
```{r}
FeaturePlot(
    immune.cells, 
    head(rownames(markers)), 
    cols.use = c("lightgrey", "blue"), 
    nCol = 3
)

FeaturePlot(immune.cells, c("Itgam"), no.legend = FALSE, cols.use = c("lightgrey", "red"), pt.size=1.2)

```

FindAllMarkers
```{r}
setwd("C:/Users/kicheol/Desktop/Projects/Andres_scRNA/Results_Seurat_new_NoCorrection/ImmuneCells")
ids = c("res.0.5","res.1","res.2","res.3")

# find all markers
rm(markers_all)
immune.cells <- SetAllIdent(immune.cells, id = ids[1])    # select resolution of tree
table(immune.cells@ident)

markers_all <- FindAllMarkers(
    object = immune.cells, 
    only.pos = TRUE, 
    min.pct = 0.1, 
    thresh.use = 0.25
)
dim(markers_all)
table(table(markers_all$gene))

write.csv(markers_all, "markers_res0.5.csv")
ImmMarkers_res0.5 <- markers_all

# find all markers
rm(markers_all)
immune.cells <- SetAllIdent(immune.cells, id = ids[2])    # select resolution of tree
table(immune.cells@ident)

markers_all <- FindAllMarkers(
    object = immune.cells, 
    only.pos = TRUE, 
    min.pct = 0.1, 
    thresh.use = 0.25
)
dim(markers_all)
table(table(markers_all$gene))

write.csv(markers_all, "markers_res1.0.csv")
ImmMarkers_res1 <- markers_all

# find all markers
rm(markers_all)
immune.cells <- SetAllIdent(immune.cells, id = ids[3])    # select resolution of tree
table(immune.cells@ident)

markers_all <- FindAllMarkers(
    object = immune.cells, 
    only.pos = TRUE, 
    min.pct = 0.1, 
    thresh.use = 0.25
)
dim(markers_all)
table(table(markers_all$gene))

write.csv(markers_all, "markers_res2.0.csv")
ImmMarkers_res2 <- markers_all

# find all markers
rm(markers_all)
immune.cells <- SetAllIdent(immune.cells, id = ids[4])    # select resolution of tree
table(immune.cells@ident)

markers_all <- FindAllMarkers(
    object = immune.cells, 
    only.pos = TRUE, 
    min.pct = 0.1, 
    thresh.use = 0.25
)
dim(markers_all)
table(table(markers_all$gene))

write.csv(markers_all, "markers_res3.0.csv")
ImmMarkers_res3 <- markers_all


#########

#
markers_all_single <- markers_all[markers_all$gene %in% names(table(markers_all$gene))[table(markers_all$gene) == 1],]
dim(markers_all_single)
table(table(markers_all_single$gene))

```



```{r}
ids = c("res.0.5","res.1","res.2","res.3")

immune.cells <- SetAllIdent(immune.cells, id = ids[4])
table(immune.cells@ident)
markers_all <- ImmMarkers_res3

#setwd("C:/Users/kicheol/Desktop/Projects/Andres_scRNA/Results_Seurat_new_NoCorrection/ImmuneCells/res0.5")


tmp_cluster <- markers_all[markers_all$cluster==0,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster00.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[37:45], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==1,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster01.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==2,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster02.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[37:45], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==3,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster03.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==4,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster04.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==5,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster05.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:34], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==6,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.1)
pdf("VlnPlot_markers_cluster06.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==7,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster07.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==8,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster08.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:32], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==9,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster09.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
dev.off()


tmp_cluster <- markers_all[markers_all$cluster==10,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster10.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==11,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster11.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==12,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster12.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
dev.off()

tmp_cluster <- markers_all[markers_all$cluster==13,]
tmp_cluster <- subset(tmp_cluster, tmp_cluster$p_val_adj < 0.05)
pdf("VlnPlot_markers_cluster13.pdf", paper="letter")
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[1:9], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[10:18], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[19:27], point.size.use = 0.2)
VlnPlot(object = immune.cells, features.plot = tmp_cluster[order(tmp_cluster$avg_logFC, decreasing=TRUE),]$gene[28:36], point.size.use = 0.2)
dev.off()

setwd("C:/Users/kicheol/Desktop/Projects/Andres_scRNA/Results_Seurat_new")
```



Plot a heatmap of genes by cluster for the top 5 marker genes per cluster
```{r}
#install.packages("dplyr")
library(dplyr)
immune.cells <- SetAllIdent(immune.cells, id = "res.1")
markers_all <- ImmMarkers_res1
markers_all_single <- markers_all[markers_all$gene %in% names(table(markers_all$gene))[table(markers_all$gene) == 1],]

top_gene <- markers_all_single %>% group_by(cluster) %>% top_n(20, avg_logFC)
dim(top_gene)
DoHeatmap(
    object = immune.cells, 
    genes.use = top_gene$gene, 
    slim.col.label = TRUE, 
    remove.key = TRUE,
    cex.row = 4
)


# violin plot for immune cell markers
setwd("C:/Users/kicheol/Desktop/Projects/Andres_scRNA/Results_Seurat_new_NoCorrection/ImmuneCells/violin_plot")

immune.cells <- SetAllIdent(immune.cells, id = "res.2")
genes <- c("Gpr34","Tmem119","Pf4","Lyz2","H2-Eb1","Cd74","P2ry12","Cx3cr1","Cxcl2","Il1a","H2-Ab1","H2-Aa","Gzma","Nkg7","Plac8","Nr4a1","Ly6i","Rhob","Sirpa","Cd9","Msrb1","Apoe","Mrc1")

for (i in 1:23){
  png(filename=paste0(genes[i],"_res2.png"), width=800, height=600)
  p <- VlnPlot(object = immune.cells, features.plot = genes[i])
  print(p)
  dev.off()
}
```


```{r}
# new function to get mean of gene expression in cluster
# Get expression of 
getGeneClusterMeans <- function(gene, cluster){
  x <- immune.cells@data[gene,]
  m <- tapply(x, ifelse(immune.cells@ident == cluster, 1, 0), mean)
  mean.in.cluster <- m[2]
  mean.out.of.cluster <- m[1]
  return(list(mean.in.cluster = mean.in.cluster, mean.out.of.cluster = mean.out.of.cluster))
}

## for sake of time only using first six (head)
means <- mapply(getGeneClusterMeans, head(markers_all[,"gene"]), head(markers_all[,"cluster"]))
means <- matrix(unlist(means), ncol = 2, byrow = T)

colnames(means) <- c("mean.in.cluster", "mean.out.of.cluster")
markers_all2 <- cbind(head(markers_all), means)

```


## Finishing up clusters.

```{r}
immune.cells <- SetAllIdent(immune.cells, id = "res.2")    # select resolution of tree
table(immune.cells@ident)
TSNEPlot(object = immune.cells, pt.size=1, do.label = T)

immune.cells <- RenameIdent(
  object = immune.cells,
  old.ident.name = c('0'),
  new.ident.name = 'Macrophage1'
)
immune.cells <- RenameIdent(
  object = immune.cells,
  old.ident.name = c('1'),
  new.ident.name = 'Microglia'
)
immune.cells <- RenameIdent(
  object = immune.cells,
  old.ident.name = c('2'),
  new.ident.name = 'Microglia'
)
immune.cells <- RenameIdent(
  object = immune.cells,
  old.ident.name = c('3'),
  new.ident.name = 'Microglia'
)
immune.cells <- RenameIdent(
  object = immune.cells,
  old.ident.name = c('4'),
  new.ident.name = 'Macrophage2'
)
immune.cells <- RenameIdent(
  object = immune.cells,
  old.ident.name = c('5'),
  new.ident.name = 'Microglia'
)
immune.cells <- RenameIdent(
  object = immune.cells,
  old.ident.name = c('6'),
  new.ident.name = 'Microglia'
)
immune.cells <- RenameIdent(
  object = immune.cells,
  old.ident.name = c('7'),
  new.ident.name = 'NK cell'
)
immune.cells <- RenameIdent(
  object = immune.cells,
  old.ident.name = c('8'),
  new.ident.name = 'Microglia'
)
immune.cells <- RenameIdent(
  object = immune.cells,
  old.ident.name = c('9'),
  new.ident.name = 'Monocyte'
)

# save final cluster (cluster or names) to metadata
immune.cells <- AddMetaData(
  object = immune.cells,
  metadata = immune.cells@ident,
  col.name = "finalcluster")

TSNEPlot(object = immune.cells, group.by="finalcluster", pt.size=1.5, do.label = T, label.size = 5)
TSNEPlot(object = immune.cells, group.by="batchid", pt.size=1.5, do.label = F)



save(immune.cells,file="Seurat_object_immune_final clusters.RData")


##### for heatmap
top_gene <- markers_all_single %>% group_by(cluster) %>% top_n(20, avg_logFC)
dim(top_gene)
DoHeatmap(
    object = immune.cells, 
    genes.use = top_gene$gene, 
    slim.col.label = TRUE, 
    remove.key = TRUE,
    cex.row = 6, group.label.rot=TRUE, group.cex = 10
)


#
TSNEPlot(object = experiment.clusters, pt.size=1, do.label = T, label.size=3.5)   # tSNE_plot_res3.5_labeled
TSNEPlot(object = experiment.clusters, pt.size=1, do.label = T, pt.shape="orig.ident", label.size=3.5)  # tSNE_plot_res3.5_labeled_shaped

TSNEPlot(object = experiment.clusters, group.by="orig.ident", pt.size=1)

TSNEPlot(object = experiment.clusters, pt.size=0.8, do.label = T)

```



### Adding in a new metadata column representing samples within clusters
```{r}
samplecluster = paste(immune.cells@meta.data$orig.ident,immune.cells@meta.data$finalcluster,sep = '-')

names(samplecluster) = rownames(immune.cells@meta.data)
head(samplecluster)

immune.cells <- AddMetaData(
  object = immune.cells,
  metadata = samplecluster,
  col.name = "samplecluster")


# set the identity to the new variable 
immune.cells <- SetAllIdent(immune.cells, id = "res.1")    # select resolution of tree
table(immune.cells@ident)
TSNEPlot(object = immune.cells, pt.size=1.2, do.label = TRUE)

cluster.comp1 <- FindMarkers(immune.cells, ident.1 = "0", ident.2= "1", test.use="wilcox")
cluster.comp2 <- FindMarkers(immune.cells, ident.1 = "0", ident.2= "1", test.use="DESeq2")
cluster.comp3 <- FindMarkers(immune.cells, ident.1 = "0", ident.2= "1", test.use="MAST")

library("VennDiagram")
library("RColorBrewer")
venn <- list("deseq2"=row.names(subset(cluster.comp2, p_val_adj<0.05)), 
             "mast"=row.names(subset(cluster.comp3, p_val_adj<0.05)),
             "wilcox"=row.names(subset(cluster.comp1, p_val_adj<0.05)))
grid.newpage()
venn.plot <- venn.diagram(venn, NULL, resolution=600, fill=brewer.pal(3, "Dark2"), col="transparent", alpha=c(0.5,0.5,0.5), 
                          main.cex = 1.5, sub.cex=2, cat.cex=1,
                          category=c("deseq2","mast","wilcox"), main="(p<0.05)")
grid.draw(venn.plot)

# significant in both wilcox and deseq2 method
geneList <- intersect(row.names(subset(cluster.comp1, p_val_adj<0.05)),row.names(subset(cluster.comp2, p_val_adj<0.05)))
DoHeatmap(immune.cells,
          use.scaled = TRUE,
          cells.use=rownames(immune.cells@meta.data)[immune.cells@meta.data$res.1 %in% c( "0", "1" )],
          genes.use = row.names(subset(cluster.comp1, p_val_adj<0.05)),
          slim.col.label = TRUE
          )

```

And last lets save all the objects in our session.
```{r}
save(list=ls(), file="clusters_seurat_object.RData")
```

Session Information
```{r}
sessionInfo()
```
